libminifi/include/properties/ConfigurationReader.h

```diff
@@ -0,0 +1,32 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <optional>
+#include <string>
+
+namespace org::apache::nifi::minifi {
+
+class ConfigurationReader {
+ public:
+  virtual bool get(const std::string& key, std::string& value) const = 0;
+  virtual bool get(const std::string& key, const std::string& alternate_key, std::string& value) const = 0;
+  virtual std::optional<std::string> get(const std::string& key) const = 0;
+};
```

Comment:
I think `std::optional<std::string> get(std::string_view key)` would cover all use cases. Alternate keys could be handled with `utils::orElse`. Consider removing this interface and replacing it with a `std::function<std::optional<std::string>(std::string_view)>` or templates with a similar callback signature, in the name of simplicity.

If you decide to keep it like this, please add a virtual destructor.

Reply:
I decided to keep the interface for now. It's more consistent with the other setters of the agent information nodes and in my opinion it's cleaner to see what that member represents.

Reply:
I mean all I'm seeing in AgentInformation is setters for random objects, and I don't think it makes a difference if that object is a custom class versus a `std::function`, a setter is a setter.

Reply:
I was referring to classes like the `AgentIdentificationProvider` or the `StateMonitor` where those classes are also implemented as separate interfaces to be later set in one of the agent information nodes. I think these adhere more to the interface segregation principle and especially in case of the FlowController where there are already way too many responsibilities they could be later extracted to separate classes and composed instead of inheriting from them.

**BUT** I rechecked this case specifically and as you mentioned we only need a single overload of the get functions and I'm not sure if this "read-only configuration" interface will be reused in the future anywhere else, so after these considerations I replaced it with a single std::function in c2544df895ed1fac5ba6d5b50e27ecea4e073c3e

Reply:
Great, thanks! One more little nitpick: is there a reason for using `const std::string&` over `std::string_view`? Unless you're working with only `std::string` objects, `std::string_view` can eliminate the need for temporary allocations when passing C strings.

Reply:
The reason was the signature of the `get` function of the `Configure` class which expects `const std::string& key` as its parameter.